proposals/0000-local-do.rst

```diff
@@ -0,0 +1,844 @@
+expressions <https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/computation-express
+ions>`_.
+
+Motivation
+------------
+
+There are many kinds of monad-like things out there:
+
+* monads
+* indexed monads
+* graded monads
+* relative monads
+* linear variants of all the above once `linear types
+  <https://github.com/ghc-proposals/ghc-proposals/pull/111>`_ are
+  implemented
+
+  * in particular, the linear IO monad from the `Linear Haskell paper
+    <https://arxiv.org/abs/1710.09756>`_ is an example of a linear
+    graded monad
+* …
+
+All of these are theoretically compatible with the do-notation. And, in fact, really want the do-notation to work for them. After all, the entire existence of the do-notation can be ascribed to: “it's really annoying to program with ``(>>=)`` all the time”.
+
+The prescribed solution is to use ``-XRebindableSyntax``. But ``-XRebindableSyntax`` is a very blunt instrument:
+
+* It affects many syntactic constructs (numerical literals, the ``if then else`` syntax, … (see the `full list <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#extension-RebindableSyntax>`_)).
+* It implies ``-XNoImplicitPrelude``
+* It uses the same rebinding for all the do-expression in an entire file.
+
+You may not want all this. For instance, with linear types ``if then else`` syntax cannot be meaningfully rebound to a function. And there is no reason why a file wouldn't have a piece code referring to a monad, one to a graded monad, and one to a linear relative monad.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal creates a new language extension ``-XQualifiedDo``.
+
+When ``-XQualifiedDo`` is activated, the syntax of the ``do`` notation is changed to
+
+::
+
+  <lexp> ⟶ [<aexp>.]do
+
+``aexp`` means that the notation before the ``.`` is parsed as a variable, unless there are parentheses.
+
+The additional expression is called the *builder* of the do-expression. The following restrictions apply to the builder and its type.
+
+* expr must **have the fully settled type** ``T``.
+* There is a type ``R`` such that normalizing ``T`` with respect to type
+  families yields a type of the form ``R T0 … Tn``.
+* ``R`` must be a datatype with precisely one constructor ``K``.
+* ``K`` must be a record constructor, defining fields with any of the following names:
+  ``(>>=)``, ``(>>)``, ``fail``, ``return``, ``<*>``, and ``<$>``.
+
+We say that an expression **has the fully settled type** ``T`` when
+
+* it is of the form ``e :: T``, or
+* it is an identifier imported from another module with type ``T``, or
+* it is of the form ``expr @ty`` where `expr` **has a fully settled type**
+  ``forall a. T``.
```

Comment:
@goldfirere, why no
```suggestion
  ``forall a. T``.
* it is of the form ``expr1 expr2`` where` `expr1`` **has fully settled type** ``T1 -> T2``
```

What breaks if you add this?

Reply:
I don't think anything breaks.

Reply:
That makes me almost like this variant again. Stuck between a pillow and a soft place, I guess.

Reply:
I agree with @facundominguez . I think that's fine. I had not realized that.